virtcontainers/qemu.go

```diff
@@ -924,8 +1012,53 @@ func (q *qemu) addDevice(devInfo interface{}, devType deviceType) error {
 
 // getSandboxConsole builds the path of the console where we can read
 // logs coming from the sandbox.
-func (q *qemu) getSandboxConsole(sandboxID string) (string, error) {
-	return utils.BuildSocketPath(runStoragePath, sandboxID, defaultConsole)
+func (q *qemu) getSandboxConsole(id string) (string, error) {
+	return utils.BuildSocketPath(RunVMStoragePath, id, consoleSocket)
+}
+
+func (q *qemu) saveSandbox() error {
+	defer func(qemu *qemu) {
+		if q.qmpMonitorCh.qmp != nil {
+			q.qmpMonitorCh.qmp.Shutdown()
+		}
+	}(q)
+
+	q.Logger().Info("save sandbox")
+
+	cfg := govmmQemu.QMPConfig{Logger: newQMPLogger()}
+
+	// Auto-closed by QMPStart().
+	disconnectCh := make(chan struct{})
+
+	qmp, _, err := govmmQemu.QMPStart(q.qmpMonitorCh.ctx, q.qmpMonitorCh.path, cfg, disconnectCh)
+	if err != nil {
+		q.Logger().WithError(err).Error("Failed to connect to QEMU instance")
+		return err
+	}
+
+	q.qmpMonitorCh.qmp = qmp
+
+	err = qmp.ExecuteQMPCapabilities(q.qmpMonitorCh.ctx)
+	if err != nil {
+		q.Logger().WithError(err).Error(qmpCapErrMsg)
+		return err
+	}
+
+	if q.config.BootToBeTemplate {
+		err = q.qmpMonitorCh.qmp.ExecSetMigrationBypassSharedMemory(q.qmpMonitorCh.ctx)
+		if err != nil {
+			q.Logger().WithError(err).Error("set migration bypass shared memory")
+			return err
+		}
+	}
+
+	err = q.qmpMonitorCh.qmp.ExecSetMigrateArguments(q.qmpMonitorCh.ctx, fmt.Sprintf("exec:cat>%s", q.config.DevicesStatePath))
```

Comment:
Are there any issues around creating the image with "version X" of qemu and then trying to use it with "version Y" of qemu? I'm thinking about build-support as well (`qemu-lite` compared with `qemu-vanilla` compared with distro-packaged `qemu`).

Reply:
In such case, the qmp command might fail. And I would classify it as a configuration mistake... It's just like the version mismatch when you upgrade kata-runtime but do not correspondingly upgrade kata-agent and end up with certain grpc failures.